std/algorithm/iteration.d

```diff
@@ -4284,3 +4284,255 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Function to compute the mean of a range, which must be an instance of
+$(D SortedRange). The range MUST be sorted in ascending order. Otherwise this
+function will give incorrect results. The elements of the range can be any type
+that can be implicitly converted to $(D real), or any user defined type that
+supports addition and division. If the element type of the range cannot be
+implicitly converted to $(D real), the $(D divisor) option must be specified
+which must be equivalent to 2 in order to correctly determine medians for even
+length ranges. If the range is empty, this function will return $(D real.nan)
+or the $(D init) value of the user type. Is $(BIGOH r.length) for forward ranges
+and $(BIGOH 1) for random access ranges.
+
+Params:
+    r = an input range with length or a forward froward range
+    divisor = optional parameter which is a value equivalent of 2
+
+Returns:
+    the median of r
+*/
+real median(R)(R r)
+    if (isInstanceOf!(SortedRange, R) &&
+        is(ElementType!R : real) &&
+        isInputRange!R &&
+        hasLength!R &&
+        !isInfinite!R)
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return median(r, real(2.0));
+}
+
+/// ditto
+real median(R)(R r)
+    if (isInstanceOf!(SortedRange, R) &&
+        is(ElementType!R : real) &&
+        isForwardRange!R &&
+        !hasLength!R &&
+        !isInfinite!R)
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return median(r, real(2.0));
+}
+
+/// ditto
+auto median(R, T)(R r, T divisor)
+    if (isInstanceOf!(SortedRange, R) &&
+        isInputRange!R &&
+        hasLength!R &&
+        is(typeof(r.front + r.front)) &&
+        is(typeof(r.front / divisor)) &&
+        !isInfinite!R)
+{
+    alias ET = Unqual!(ElementType!R);
+
+    if (r.empty)
+    {
+        return ET.init;
+    }
+
+    if (r.length == 0)
+    {
+        return r.front;
+    }
+
+    static if (isRandomAccessRange!R)
+    {
+        if (r.length % 2 == 0)
+        {
+            return (r[(r.length / 2) - 1] + r[(r.length / 2)]) / divisor;
+        }
+        else
+        {
+            return r[(r.length - 1) / 2];
+        }
+    }
+    else
+    {
+        if (r.length % 2 == 0)
+        {
+            r = r.drop((r.length / 2) - 1);
+            ET prev_item = r.front;
+            r.popFront;
+
+            return (r.front + prev_item) / divisor;
+        }
+        else
+        {
+            r.drop((r.length - 1) / 2);
+
+            return r.front;
+        }
+    }
+
+    assert(0);
+}
+
+/// ditto
+auto median(R, T)(R r, T divisor)
+    if (isInstanceOf!(SortedRange, R) &&
+        isForwardRange!R &&
+        is(typeof(r.front + r.front)) &&
+        is(typeof(r.front / r.front)) &&
+        !hasLength!R &&
+        !isInfinite!R)
+{
+    alias ET = Unqual!(ElementType!R);
+
+    if (r.empty)
+    {
+        return ET.init;
+    }
+
+    size_t length = r.save.walkLength;
```

Comment:
The only diff with not RA but has length and this code is how we compute length. Just add a helper that takes a range and a precomputed length and use it in both cases. In case of hasLength!R == True you pass `medianImpl(r, r.length)` and otherwise `medianImpl(r, r.save.walkLength)`